array_api_tests/test_signatures.py

```diff
@@ -1,282 +1,243 @@
-    else:
-        assert hasattr(mod, name), f"{mod_name} is missing the {function_category(name)} function {name}()"
-
-@pytest.mark.parametrize('name', params)
-def test_function_positional_args(name):
-    # Note: We can't actually test that positional arguments are
-    # positional-only, as that would require knowing the argument name and
-    # checking that it can't be used as a keyword argument. But argument name
-    # inspection does not work for most array library functions that are not
-    # written in pure Python (e.g., it won't work for numpy ufuncs).
-
-    if extension_module(name):
-        return
-
-    dtype = None
-    if (name.startswith('__i') and name not in ['__int__', '__invert__', '__index__']
-        or name.startswith('__r') and name != '__rshift__'):
-        n = f'__{name[3:]}'
-    else:
-        n = name
-    in_dtypes = dh.func_in_dtypes.get(n, dh.float_dtypes)
-    if bool in in_dtypes:
-        dtype = bool
-    elif all(d in in_dtypes for d in dh.all_int_dtypes):
-        dtype = int64
-
-    if array_method(name):
-        if name == '__bool__':
-            _mod = ones((), dtype=bool)
-        elif name in ['__int__', '__index__']:
-            _mod = ones((), dtype=int64)
-        elif name == '__float__':
-            _mod = ones((), dtype=float64)
+"""
+from inspect import Parameter, Signature, signature
+from types import FunctionType
+from typing import Any, Callable, Dict, List, Literal, get_args
+
+import pytest
+from hypothesis import given, note, settings
+from hypothesis import strategies as st
+from hypothesis.strategies import DataObject
+
+from . import dtype_helpers as dh
+from . import hypothesis_helpers as hh
+from . import xps
+from ._array_module import _UndefinedStub
+from ._array_module import mod as xp
+from .stubs import array_methods, category_to_funcs, extension_to_funcs
+from .typing import Array, DataType
+
+pytestmark = pytest.mark.ci
+
+ParameterKind = Literal[
+    Parameter.POSITIONAL_ONLY,
+    Parameter.VAR_POSITIONAL,
+    Parameter.POSITIONAL_OR_KEYWORD,
+    Parameter.KEYWORD_ONLY,
+    Parameter.VAR_KEYWORD,
+]
+ALL_KINDS = get_args(ParameterKind)
+VAR_KINDS = (Parameter.VAR_POSITIONAL, Parameter.VAR_KEYWORD)
+kind_to_str: Dict[ParameterKind, str] = {
+    Parameter.POSITIONAL_OR_KEYWORD: "pos or kw argument",
+    Parameter.POSITIONAL_ONLY: "pos-only argument",
+    Parameter.KEYWORD_ONLY: "keyword-only argument",
+    Parameter.VAR_POSITIONAL: "star-args (i.e. *args) argument",
+    Parameter.VAR_KEYWORD: "star-kwargs (i.e. **kwargs) argument",
+}
+
+
+def _test_inspectable_func(sig: Signature, stub_sig: Signature):
+    params = list(sig.parameters.values())
+    stub_params = list(stub_sig.parameters.values())
+    # We're not interested if the array module has additional arguments, so we
+    # only iterate through the arguments listed in the spec.
+    for i, stub_param in enumerate(stub_params):
```

Comment:
Would this catch functions that have too many positional-only arguments?

Reply:
Yep! We're iterating over the parameters from the `signatures/` stubs and checking them one-by-one, which covers this (added a test for this too).

Reply:
Yep! We're iterating over the parameters from the `signatures/` stubs and checking them one-by-one, which covers this (added a test for this too).